### PR TITLE
fix: Bump Kubernetes 1.29.2

### LINKED ISSUE #16473: More friendly error message when label value is nil to improve userability
# Summary

In `repo-server`, when label is nil, there manifest generation would be failed and will got below error message in UI:
```text
Failed to load target state: failed to generate manifest for source 1 of 1: rpc error: code = Unknown desc = failed to set app instance tracking info on manifest: failed to set app instance label: failed to get labels from target object apps/v1, Kind=Deployment /rgo-product-ingestion-service: .metadata.labels accessor error: contains non-string key in the map: <nil> is of the type <nil>, expected string
```

The error message is not friendly, could we point out which label is nil in error message? Like 
```text
... the value of label <key1> is nil from target object apps/v1, Kind=Deployment /rgo-product-ingestion-service...
```

# Motivation
Imporve userability

# Proposal

Change the error message to 

```text
... the value of label <key1> is nil from target object apps/v1, Kind=Deployment /rgo-product-ingestion-service...
```

### LINKED ISSUE #17490: Update dependencies to Kubernetes.
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

Now, ArgoCD depends some Kubernetes libraries that the version is 0.26.1. And recent version of these  are released up to 0.29.2, so if we use some package of ArgoCD (in my case I use CRD) and Kubernetes libraries later 0.28.0, we can't resolve dependencies and fail to build, due to some interfaces are changed. Then if we can settle this problem, it needs to update the version of Kubernetes that gitops-engine depends in advance. If you have free time to consider this, please watch it some changes. Thanks.

### PR DESCRIPTION
**Pull request by sivchari**
close #17490 

closes https://github.com/argoproj/argo-cd/issues/16473

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -280,18 +281,20 @@ require (
...
 	sigs.k8s.io/json v0.0.0-20221116044647-bc3834ca7abd // indirect
-	sigs.k8s.io/kustomize/api v0.12.1 // indirect
-	sigs.k8s.io/kustomize/kyaml v0.13.9 // indirect
+	sigs.k8s.io/kustomize/api v0.13.5-0.20230601165947-6ce0bf390ce3 // indirect
+	sigs.k8s.io/kustomize/kyaml v0.14.3-0.20230601165947-6ce0bf390ce3 // indirect
 )
 
 replace (
+
+	github.com/argoproj/gitops-engine => github.com/sivchari/gitops-engine v0.0.0-20240228144537-7c3b9522429b
```
Comment (sivchari) [inline go.mod:297]: I temporally use [some of  my changes](https://github.com/argoproj/gitops-engine/pull/566) here. So I need to replace it after this pull-request is merged.

**Comment by sivchari** [comment]
Both [this commit](https://github.com/argoproj/argo-cd/pull/17491/commits/6439add760398c6350c1f4c550f8a82de802f786) and [one more commit](https://github.com/argoproj/argo-cd/pull/17491/commits/c2bb88a7c538e5bea2737c874e4bbceb61c1d253) is to fix some errors by lint on CI, due to updating libraries.

**Comment by blakepettersson** [comment]
Hello @sivchari,

There seems to be some codegen issues:

 * Can you modify `hack/update-codegen.sh` to [replace `generate-groups.sh` with `kube_codegen.sh`](https://github.com/argoproj/argo-cd/blob/4abc9929281296122db0eace1607583187dd9944/hack/update-codegen.sh#L30-L33)
 * In `hack/install-codegen-go-tools.sh` can you remove [this line](https://github.com/argoproj/argo-cd/blob/4abc9929281296122db0eace1607583187dd9944/hack/installers/install-codegen-go-tools.sh#L44-L45) (line 45)?

Then run `make codegen` again - hopefully that should get you past that hurdle at least

**Comment by sivchari** [comment]
Hi, @blakepettersson 

CI has still been failed and the log is below.

```
+ INT_APIS_PKG=
++ dirname /tmp/generate-groups.sh
+ exec /tmp/generate-internal-groups.sh deepcopy,client,informer,lister github.com/argoproj/argo-cd/v2/pkg/client '' github.com/argoproj/argo-cd/v2/pkg/apis application:v1alpha1 --go-header-file /home/runner/go/src/github.com/argoproj/argo-cd/hack/custom-boilerplate.go.txt --output-base /home/runner/go/src
/tmp/generate-groups.sh: line 66: /tmp/generate-internal-groups.sh: No such file or directory
make: *** [Makefile:219: clientgen] Error 127
```

What should I do ? Thanks.

**Comment by sivchari** [comment]
Hello @blakepettersson 
Should I contains these changes which you mentioned into commit ? I had already done two steps that you told me.

**Comment by sivchari** [comment]
Can I extend the timeout ?

**Comment by blakepettersson** [comment]
@sivchari the timeout is already exceeding ~2x of what the e2e tests normally take - there is a panic in `e2e.TestAppSyncWrongVersion`, either in the test itself or in the underlying code which is being tested which needs to be addressed

File: Makefile [hunk 2]
```diff
@@ -224,7 +224,7 @@ clidocsgen:
 
 
 .PHONY: codegen-local
-codegen-local: mod-vendor-local gogen protogen clientgen openapigen clidocsgen manifests-local notification-docs notification-catalog
+codegen-local: mod-vendor-local gogen protogen clientgen clidocsgen manifests-local notification-docs notification-catalog
```
Comment (blakepettersson) [inline Makefile:227]: I don't think this should change actually; `openapigen` should still run.

**Comment by sivchari** [comment]
> the timeout is already exceeding ~2x of what the e2e tests normally take - there is a panic in e2e.TestAppSyncWrongVersion, either in the test itself or in the underlying code which is being tested which needs to be addressed

@blakepettersson 
I got it. I'd investigate, thanks.

**Comment by sivchari** [comment]
スクリーンショット 2024-04-16 12 10 46

golangci-lint reports uncorrect errors.

**Comment by sivchari** [comment]
> error when replacing \"/dev/shm/180188768\": Deployment.apps \"guestbook-ui\" is invalid: spec.selector: Invalid value: v1.LabelSelector{MatchLabels:map[string]string{\"app\":\"guestbook-ui\", \"env\":\"e2e\"}, MatchExpressions:[]v1.LabelSelectorRequirement(nil)}: field is immutable\

I fixed the cache options which namespace should be watched.
If these tests will be passed, but same log will be displayed, we can assume this bug is another issue.
Thus, I wait the result of tests.
@crenshaw-dev 
What do you think ?

**Comment by sivchari** [comment]
@crenshaw-dev 

https://github.com/argoproj/argo-cd/issues/12439

This error seems to be happened since the above.

**Comment by sivchari** [comment]
I guess this error depends on version of kubectl.

**Comment by sivchari** [comment]
OK, I solved it.
The dependency which replaces gitops-engine didn't include [this commit](https://github.com/argoproj/gitops-engine/commit/fbecbb86e41254a75a59943b5eb43ed55d21cdb9). Then, ArgoCD couldn't replace forcibly, probably because of kubectl is upgraded. So I made ArgoCD use gitops-engine including latest commit.
It worked in local environmentental. Thus checks might have been completed later.

**Comment by sivchari** [comment]
@backjo @crenshaw-dev 
All test is passed 👍 
I've used my pull-request which is submitet to gitops-engine yet. So, when it is merged, this pull-request will be ready to merge.
Thanks.

File: Makefile (hunk 2 above)
Comment (blakepettersson) [reply Makefile:227]: @sivchari did you see my comment above? ☝️

Comment (sivchari) [reply Makefile:227]: @blakepettersson 
Oh, I've completely overlooked.
I fixed it.

**Comment by blakepettersson** [comment]
@sivchari  hmm, perhaps `openapigen` should be omitted after all in `make codegen`, but I don't think so 🤔

File: Makefile (hunk 2 above)
Comment (sivchari) [reply Makefile]: @blakepettersson 

openapigen has already been deleted. So, if openapigen is remained, the CI has never been passed.

ref: https://github.com/argoproj/argo-cd/pull/17491#issuecomment-2043643951

**Comment by sivchari** [comment]
@blakepettersson 
I don't have better way to solve this issue on my laptop, but I agree with your opinion.
What should I do ??

**Comment by sivchari** [comment]
@blakepettersson 
If you feel free to take over my commit, could you do the part of codegen ? Presumably, it's problem with my environmental (of course, we should deal with this issue in the future).

File: hack/installers/install-codegen-go-tools.sh [hunk 3]
```diff
@@ -41,9 +41,6 @@ go_mod_install k8s.io/code-generator/cmd/defaulter-gen
 go_mod_install k8s.io/code-generator/cmd/informer-gen
 go_mod_install k8s.io/code-generator/cmd/lister-gen
 
-# We still install openapi-gen from go.mod since upstream does not utilize release tags
-go_mod_install k8s.io/kube-openapi/cmd/openapi-gen
```
Comment (crenshaw-dev) [inline hack/installers/install-codegen-go-tools.sh:45]: I think this removal caused the codegen error. Has this binary been moved/renamed in 1.29.2?

Comment (crenshaw-dev) [reply hack/installers/install-codegen-go-tools.sh]: re-added, let's see how it behaves

File: util/clusterauth/clusterauth.go [hunk 4]
```diff
@@ -258,24 +258,11 @@ func GetServiceAccountBearerToken(clientset kubernetes.Interface, ns string, sa
...
-		if len(serviceAccount.Secrets) == 0 {
-			return false, nil
-		}
-		return true, nil
-	})
-	if err != nil && err != wait.ErrWaitTimeout {
-		return "", fmt.Errorf("failed to get serviceaccount token secret: %w", err)
+	serviceAccount, err := clientset.CoreV1().ServiceAccounts(ns).Get(context.Background(), sa, metav1.GetOptions{})
+	if err != nil {
+		return "", fmt.Errorf("failed to get serviceaccount %q: %w", sa, err)
```
Comment (crenshaw-dev) [inline util/clusterauth/clusterauth.go:263]: I'm surprised to see the removal of the poll mechanism here. What's the intent?

File: util/kube/kube_test.go [hunk 5]
```diff
@@ -192,7 +192,7 @@ func TestSetAppInstanceAnnotationWithInvalidData(t *testing.T) {
 	assert.Nil(t, err)
 	err = SetAppInstanceAnnotation(&obj, common.LabelKeyAppInstance, "my-app")
 	assert.Error(t, err)
-	assert.Equal(t, "failed to get annotations from target object /v1, Kind=Service /my-service: .metadata.annotations accessor error: contains non-string key in the map: <nil> is of the type <nil>, expected string", err.Error())
+	assert.Equal(t, "failed to get annotations from target object /v1, Kind=Service /my-service: .metadata.annotations accessor error: contains non-string value in the map under key \"invalid-annotation\": <nil> is of the type <nil>, expected string", err.Error())
```
Comment (crenshaw-dev) [inline util/kube/kube_test.go:195]: Oh heck yes, my upstream change finally landed! https://github.com/kubernetes/kubernetes/pull/118989

File: util/clusterauth/clusterauth.go (hunk 4 above)
Comment (crenshaw-dev) [reply util/clusterauth/clusterauth.go:263]: Ah interesting. Okay, wanna go ahead and revert it? I think we're past the codegen issues.

Comment (sivchari) [reply util/clusterauth/clusterauth.go:263]: Okay, I have a bit question. Is the problem related with codegen issue?
If so, I agree to revert it.
If you feel free to revert it, could you do it ?

Comment (sivchari) [reply util/clusterauth/clusterauth.go]: I reverted it.

Comment (crenshaw-dev) [reply util/clusterauth/clusterauth.go]: Nah, should be unrelated. Thanks!

**Comment by sivchari** [comment]
Some tests are failed due to rate limit too.

**Comment by crenshaw-dev** [comment]
Looks possibly related:

```
    deployment_test.go:322: 
        	Error Trace:	/home/runner/work/argo-cd/argo-cd/test/e2e/deployment_test.go:322
        	            				/home/runner/work/argo-cd/argo-cd/test/e2e/deployment_test.go:173
        	Error:      	Expected nil, but got: &fmt.wrapError{msg:"failed to create secret for serviceaccount \"\": an empty namespace may not be set during creation", err:(*errors.errorString)(0xc00045f410)}
```

**Comment by sivchari** [comment]
rate limit error, too.

**Comment by sivchari** [comment]
@crenshaw-dev 
Okay, I solved it. This problem was happened due to that we don't exit the loop if service account doesn't have secrets.
The new function wait until returning value is true or non-nil error. Since we'll create new secret when service account doesn't have secret, I modify the logic that exit the loop even though service account doesn't have secret.

**Comment by sivchari** [comment]
@crenshaw-dev @blakepettersson 
All test are passed 😄

**Comment by blakepettersson** [comment]
Great work @sivchari!

File: cmd/argocd-applicationset-controller/commands/applicationset_controller.go [hunk 6]
```diff
@@ -113,15 +116,28 @@ func NewCommand() *cobra.Command {
...
-				Scheme:                 scheme,
-				MetricsBindAddress:     metricsAddr,
-				Namespace:              watchedNamespace,
+				Scheme: scheme,
+				Metrics: metricsserver.Options{
+					BindAddress: metricsAddr,
+				},
+				Cache:                  cacheOpt,
 				HealthProbeBindAddress: probeBindAddr,
-				Port:                   9443,
```
Comment (crenshaw-dev) [inline cmd/argocd-applicationset-controller/commands/applicationset_controller.go:121]: Is this option intentionally dropped? Does it default to 9443?

File: common/common.go [hunk 7]
```diff
@@ -300,7 +300,7 @@ const (
 
 // Constants used by util/clusterauth package
 const (
-	ClusterAuthRequestTimeout = 10 * time.Second
+	ClusterAuthRequestTimeout = 30 * time.Second
```
Comment (crenshaw-dev) [inline common/common.go:303]: Can this go back to 10?

File: util/clusterauth/clusterauth.go [hunk 8]
```diff
@@ -430,7 +427,7 @@ func GenerateNewClusterManagerSecret(clientset kubernetes.Interface, claims *Ser
 		return true, nil
 	})
 	if err != nil {
-		return nil, fmt.Errorf("Timed out waiting for secret to generate new token")
+		return nil, fmt.Errorf("Timed out waiting for secret to generate new token: err %w", err)
```
Comment (crenshaw-dev) [inline util/clusterauth/clusterauth.go:430]:
```suggestion
		return nil, fmt.Errorf("Timed out waiting for secret to generate new token: %w", err)
```

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]
Legendary work, @sivchari! Just 3 small things, then I think we'll be good to go.

File: cmd/argocd-applicationset-controller/commands/applicationset_controller.go (hunk 6 above)
Comment (sivchari) [reply cmd/argocd-applicationset-controller/commands/applicationset_controller.go:121]: Yes.
https://github.com/kubernetes-sigs/controller-runtime/blob/479b723944e34ae42c9911fe01228ff34eb5ca81/pkg/webhook/server.go#L39
The port of 9443 became by default

File: common/common.go (hunk 7 above)
Comment (sivchari) [reply common/common.go:303]: Thanks.

**Comment by sivchari** [comment]
@crenshaw-dev 
I fixed and replied these nits.

**Comment by sivchari** [comment]
After [it](https://github.com/argoproj/gitops-engine/pull/566) is merged, we can be ready to merge this PR.

**Comment by crenshaw-dev** [comment]
Merged!

**Comment by sivchari** [comment]
@crenshaw-dev 
I placed it.

**Review by crenshaw-dev — APPROVED** [review]
Brilliant, thanks!

**Comment by morremeyer** [comment]
Can this be cherry-picked for v2.11.2? It would likely resolve https://github.com/argoproj/argo-cd/issues/15176.

**Comment by morremeyer** [comment]
Looks like the cherry-pick bot didn't pick this one up 🤔

**Comment by asafhm** [comment]
@morremeyer is right
@pasha-codefresh - can we try again please? 🙏🏻

**Comment by pasha-codefresh** [comment]
I will cherry-pick it today later, not sure why bot is not picking it

**Comment by blakepettersson** [comment]
Do we really want to do such a big bump of client-go for a patch release?

**Comment by morremeyer** [comment]
It’s a backward compatible fix of an issue, which semver defines as patch bump.

I don’t see any issue in including it, it should resolve an issue that makes ArgoCD unusable for some cases - which for me makes it clear that it should be released asap.

Disclaimer: I’m a huge proponent of releasing every fix and feature as soon as merged to the default/release branch, which is generally not the policy of Argo, so take my opinion with a grain of salt

**Comment by pasha-codefresh** [comment]
After today's ArgoCD contribution meeting, we decided not to cherry-pick these changes into version 2.11 as a patch. Instead, we are planning to release version 2.12 RC1 in two weeks, where this change will be included.

**Comment by asafhm** [comment]
@pasha-codefresh In my opinion, this fix would be very helpful if it was delivered quickly.

I'd love to make my case here quick so I can maybe try changing your mind...
The following combination is probably what makes me suffer from this fix's absence:

- I use a Kubernetes version above 1.26 (1.30 is already out there for early adopters.)
- I have deployments/statefulsets with `topologySpreadConstraints` configured with a field that was introduced as beta in Kubernetes 1.27 (the `matchLabelKeys` field, which may be in beta, though sought after and incredibly helpful)
- I have `ignoreDifferences` settings that operate on said deployments/statefulsets (these are probably the most popular resources, and I assume a lot of users try to setup `ignoreDifferences` on them)

**Comment by Whisper40** [comment]
Hi, thanks for the contribution, i just lost 2 days on my side trying to understand what was wrong, but now i understand that the fix has not been cherrypick :)

**Pull request merged**